In [16]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [17]:
# import datasets
df_train = pd.read_csv('../assets/train.csv')
df_spray = pd.read_csv('../assets/spray.csv')
df_weather = pd.read_csv('../assets/weather.csv')
map_data = np.loadtxt('../assets/mapdata_copyright_openstreetmap_contributors.txt')

Every year from late-May to early-October, public health workers in Chicago setup mosquito traps scattered across the city. Every week from Monday through Wednesday, these traps collect mosquitos, and the mosquitos are tested for the presence of West Nile virus before the end of the week. The test results include the number of mosquitos, the mosquitos species, and whether or not West Nile virus is present in the cohort. 

These test results are organized in such a way that when the number of mosquitos exceed 50, they are split into another record (another row in the dataset), such that the number of mosquitos are capped at 50. 

The location of the traps are described by the block number and street name. For your convenience, we have mapped these attributes into Longitude and Latitude in the dataset. Please note that these are derived locations. For example, Block=79, and Street= "W FOSTER AVE" gives us an approximate address of "7900 W FOSTER AVE, Chicago, IL", which translates to (41.974089,-87.824812) on the map.

Some traps are "satellite traps". These are traps that are set up near (usually within 6 blocks) an established trap to enhance surveillance efforts. Satellite traps are postfixed with letters. For example, T220A is a satellite trap to T220. 

Please note that not all the locations are tested at all times. Also, records exist only when a particular species of mosquitos is found at a certain trap at a certain time. In the test set, we ask you for all combinations/permutations of possible predictions and are only scoring the observed ones.

- Id: the id of the record
- Date: date that the WNV test is performed
- Address: approximate address of the location of trap. This is used to send to the GeoCoder. 
- Species: the species of mosquitos
- Block: block number of address
- Street: street name
- Trap: Id of the trap
- AddressNumberAndStreet: approximate address returned from GeoCoder
- Latitude, Longitude: Latitude and Longitude returned from GeoCoder
- AddressAccuracy: accuracy returned from GeoCoder
- NumMosquitos: number of mosquitoes caught in this trap
- WnvPresent: whether West Nile Virus was present in these mosquitos. 1 means WNV is present, and 0 means not present. 

In [18]:
df_train.head()

Date                                            Address  \
0  2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1  2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2  2007-05-29  6200 North Mandell Avenue, Chicago, IL 60646, USA   
3  2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   
4  2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2          CULEX RESTUANS     62    N MANDELL AVE  T007   
3  CULEX PIPIENS/RESTUANS     79     W FOSTER AVE  T015   
4          CULEX RESTUANS     79     W FOSTER AVE  T015   

              AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
1  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
2   6200  N MANDELL AVE, Chicago, IL  41.994991 -87.769279                9   
3    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   
4    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   

   NumMosquitos  WnvPresent  
0             1           0  
1             1           0  
2             1           0  
3             1           0  
4             4           0

In [21]:
df_train.isnull().sum()

Date                      0
Address                   0
Species                   0
Block                     0
Street                    0
Trap                      0
AddressNumberAndStreet    0
Latitude                  0
Longitude                 0
AddressAccuracy           0
NumMosquitos              0
WnvPresent                0
dtype: int64

The City of Chicago also does spraying to kill mosquitos. You are given the GIS data for their spray efforts in 2011 and 2013. Spraying can reduce the number of mosquitos in the area, and therefore might eliminate the appearance of West Nile virus. 

In [19]:
df_spray.head()

Date        Time   Latitude  Longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [27]:
df_spray.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14835 entries, 0 to 14834
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       14835 non-null  object 
 1   Time       14251 non-null  object 
 2   Latitude   14835 non-null  float64
 3   Longitude  14835 non-null  float64
dtypes: float64(2), object(2)
memory usage: 463.7+ KB


In [40]:
df_spray.isnull().sum()

Date           0
Time         584
Latitude       0
Longitude      0
datetime     584
dtype: int64

In [32]:
# combine date and time into a datetime format
df_spray['datetime'] = df_spray['Date'] + ' ' + df_spray['Time']
df_spray['datetime'] = pd.to_datetime(df_spray['datetime'])

In [34]:
df_spray.tail(20)

Date        Time   Latitude  Longitude            datetime
14815  2013-09-05  8:30:21 PM  42.007355 -87.813768 2013-09-05 20:30:21
14816  2013-09-05  8:30:31 PM  42.007818 -87.813775 2013-09-05 20:30:31
14817  2013-09-05  8:30:41 PM  42.008387 -87.813745 2013-09-05 20:30:41
14818  2013-09-05  8:30:51 PM  42.008933 -87.813720 2013-09-05 20:30:51
14819  2013-09-05  8:31:01 PM  42.009487 -87.813683 2013-09-05 20:31:01
14820  2013-09-05  8:31:11 PM  42.010087 -87.813695 2013-09-05 20:31:11
14821  2013-09-05  8:31:21 PM  42.010682 -87.813713 2013-09-05 20:31:21
14822  2013-09-05  8:31:31 PM  42.011213 -87.813688 2013-09-05 20:31:31
14823  2013-09-05  8:31:51 PM  42.011340 -87.812347 2013-09-05 20:31:51
14824  2013-09-05  8:33:11 PM  42.010135 -87.812330 2013-09-05 20:33:11
14825  2013-09-05  8:33:21 PM  42.009693 -87.812338 2013-09-05 20:33:21
14826  2013-09-05  8:33:31 PM  42.009125 -87.812338 2013-09-05 20:33:31
14827  2013-09-05  8:33:41 PM  42.008507 -87.812357 2013-09-05 20:33:41
14828  2013-09-05  8:33:51 PM  42.007897 -87.812335 2013-09-05 20:33:51
14829  2013-09-05  8:34:01 PM  42.007218 -87.812338 2013-09-05 20:34:01
14830  2013-09-05  8:34:11 PM  42.006587 -87.812355 2013-09-05 20:34:11
14831  2013-09-05  8:35:01 PM  42.006192 -87.816015 2013-09-05 20:35:01
14832  2013-09-05  8:35:21 PM  42.006022 -87.817392 2013-09-05 20:35:21
14833  2013-09-05  8:35:31 PM  42.005453 -87.817423 2013-09-05 20:35:31
14834  2013-09-05  8:35:41 PM  42.004805 -87.817460 2013-09-05 20:35:41

In [41]:
len(df_spray[df_spray['Time'].isnull()])

584

In [42]:
df_spray.groupby("Date").Time.count()

Date
2011-08-29      95
2011-09-07    1530
2013-07-17    2202
2013-07-25    1607
2013-08-08    1195
2013-08-15    2668
2013-08-16     141
2013-08-22    1587
2013-08-29    2302
2013-09-05     924
Name: Time, dtype: int64

It is believed that hot and dry conditions are more favorable for West Nile virus than cold and wet. We provide you with the dataset from NOAA of the weather conditions of 2007 to 2014, during the months of the tests. 

Station 1: CHICAGO O'HARE INTERNATIONAL AIRPORT Lat: 41.995 Lon: -87.933 Elev: 662 ft. above sea level
Station 2: CHICAGO MIDWAY INTL ARPT Lat: 41.786 Lon: -87.752 Elev: 612 ft. above sea level

In [39]:
df_weather.head().transpose()

0           1           2           3           4
Station               1           2           1           2           1
Date         2007-05-01  2007-05-01  2007-05-02  2007-05-02  2007-05-03
Tmax                 83          84          59          60          66
Tmin                 50          52          42          43          46
Tavg                 67          68          51          52          56
Depart               14           M          -3           M           2
DewPoint             51          51          42          42          40
WetBulb              56          57          47          47          48
Heat                  0           0          14          13           9
Cool                  2           3           0           0           0
Sunrise            0448           -        0447           -        0446
Sunset             1849           -        1850           -        1851
CodeSum                                      BR       BR HZ            
Depth                 0           M           0           M           0
Water1                M           M           M           M           M
SnowFall            0.0           M         0.0           M         0.0
PrecipTotal        0.00        0.00        0.00        0.00        0.00
StnPressure       29.10       29.18       29.38       29.44       29.39
SeaLevel          29.82       29.82       30.09       30.08       30.12
ResultSpeed         1.7         2.7        13.0        13.3        11.7
ResultDir            27          25           4           2           7
AvgSpeed            9.2         9.6        13.4        13.4        11.9

In [38]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2944 entries, 0 to 2943
Data columns (total 22 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Station      2944 non-null   int64  
 1   Date         2944 non-null   object 
 2   Tmax         2944 non-null   int64  
 3   Tmin         2944 non-null   int64  
 4   Tavg         2944 non-null   object 
 5   Depart       2944 non-null   object 
 6   DewPoint     2944 non-null   int64  
 7   WetBulb      2944 non-null   object 
 8   Heat         2944 non-null   object 
 9   Cool         2944 non-null   object 
 10  Sunrise      2944 non-null   object 
 11  Sunset       2944 non-null   object 
 12  CodeSum      2944 non-null   object 
 13  Depth        2944 non-null   object 
 14  Water1       2944 non-null   object 
 15  SnowFall     2944 non-null   object 
 16  PrecipTotal  2944 non-null   object 
 17  StnPressure  2944 non-null   object 
 18  SeaLevel     2944 non-null   object 
 19  Result